core/domain/commit_commands_domain.py

```diff
@@ -0,0 +1,317 @@
+            required keys or there are additional keys apart from the
+            required and optional keys.
+    """
+
+    missing_keys = [key for key in required_keys if key not in actual_keys]
+
+    extra_keys = [key for key in actual_keys if key not in (
+        required_keys + optional_keys)]
+
+    error_msg = ''
+    if missing_keys:
+        error_msg = error_msg + (
+            'The following required keys are missing: %s' % (
+                (',').join(missing_keys)))
+        if extra_keys:
+            error_msg = error_msg + ', '
+
+    if extra_keys:
+        error_msg = error_msg + 'The following extra keys are present: %s' % (
+            (',').join(extra_keys))
+
+    if missing_keys + extra_keys:
+        raise utils.ValidationError(error_msg)
+
+
+class BaseCommitCmd(object):
+    """Domain object for commit commands of a storage model.
+
+    Attributes:
+        name: str. The command name.
+        parameters: dict. The command parameters.
+    """
+
+    def __init__(self, name, parameters):
+        """Constructs a CommitCmd domain object.
+
+        Args:
+            name: str. The command name.
+            parameters: dict. The command parameters.
+        """
+        self.name = name
+        self.parameters = parameters
+        self.command_list = []
+
+    def validate(self):
+        """Checks that all parameters of a command are valid for the given
+        command.
+        """
+
+        required_keys = []
+        optional_keys = []
+        for cmd in self.command_list:
+            if cmd['name'] == self.name:
+                required_keys = cmd['required_keys']
+                optional_keys = cmd['optional_keys']
+                break
+        validate_dict_keys(self.parameters.keys(), required_keys, optional_keys)
+
+
+class ExplorationCommitCmd(BaseCommitCmd):
+    """Domain object for commit commands of a exploration.
+
+    Attributes:
+        name: str. The command name.
+        parameters: dict. The command parameters.
+    """
+
+    def __init__(self, name, parameters):
+        super(ExplorationCommitCmd, self).__init__(name, parameters)
+        self.command_list = [{
+            'name': exp_domain.CMD_CREATE_NEW,
+            'required_keys': ['category', 'title'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_ADD_STATE,
+            'required_keys': ['state_name'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_DELETE_STATE,
+            'required_keys': ['state_name'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_RENAME_STATE,
+            'required_keys': ['new_state_name', 'old_state_name'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_EDIT_STATE_PROPERTY,
+            'required_keys': ['property_name', 'state_name', 'new_value'],
+            'optional_keys': ['old_value']
```

Comment:
The docstring in `exp_domain.ExplorationChange` specifies the keys as optional. Same is the case for `collection_domain.CollectionChange`. But in `story_domain.StoryChange` no key is specified as optional. 

Further, the old_values are used in StoryChange by doing a `copy.deepcopy` as follows:

`self.old_value = copy.deepcopy(change_dict['old_value'])`

In case of exploration and collection since the old_value is optional, we use a get method on the change_dict instead.

`self.old_value = change_dict.get('old_value')`

Reply:
Ah, that's a very good catch, and thank you for reading the docstrings. What you've done sounds good. Please add a comment to explain, something like: "Per the docstring in exp_domain.ExplorationChange, the old_value key is optional."

Reply:
Updated the complete structure of commit commands domain.